viamupnp/find.go

```diff
@@ -0,0 +1,165 @@
+)
+
+// DeviceQuery specifics a query for a network device.
+type DeviceQuery struct {
+	ModelName    string `json:"model_name"`
+	Manufacturer string `json:"manufacturer"`
+	SerialNumber string `json:"serial_number"`
+	Network      string `json:"network"`
+}
+
+// UPNPDevice is a UPNPDevice device.
+type UPNPDevice struct {
+	Service ssdp.Service
+	Desc    *deviceDesc
+}
+
+// FindHost looks for a host matching the query, returns just the host/ip (no port).
+func FindHost(ctx context.Context, logger logging.Logger, query DeviceQuery) (string, error) {
+	all, err := findAll(ctx, logger, query.Network)
+	if err != nil {
+		return "", err
+	}
+
+	for _, a := range all {
+		if a.Matches(query) {
+			u, err := url.Parse(a.Service.Location)
+			if err != nil {
+				// should be impossible
+				logger.Warnf("invalid location %s", a.Service.Location)
+				continue
+			}
+			return strings.Split(u.Host, ":")[0], nil
+		}
+	}
+
+	return "", fmt.Errorf("no match found for query: %v", query)
+}
+
+func matches(query string, s string) bool {
+	if query == s {
+		return true
+	}
+
+	if strings.HasSuffix(query, ".*") {
+		query = query[0 : len(query)-2]
+		return strings.HasPrefix(s, query)
+	}
+
+	return false
+}
+
+// Matches returns if the UPNPDevice matches the query.
+func (pc *UPNPDevice) Matches(query DeviceQuery) bool {
+	if query.ModelName != "" && !matches(query.ModelName, pc.Desc.Device.ModelName) {
+		return false
+	}
+
+	if query.Manufacturer != "" && !matches(query.Manufacturer, pc.Desc.Device.Manufacturer) {
+		return false
+	}
+
+	if query.SerialNumber != "" && !matches(query.SerialNumber, pc.Desc.Device.SerialNumber) {
+		return false
+	}
+
+	return true
+}
+
+// FindAllTestKeyStruct - for testing.
+type FindAllTestKeyStruct string
+
+// FindAllTestKey - for testing.
+const FindAllTestKey = FindAllTestKeyStruct("findAllTestKey1231231231231")
+
+func findAll(ctx context.Context, logger logging.Logger, network string) ([]UPNPDevice, error) {
+	all, ok := ctx.Value(FindAllTestKey).([]UPNPDevice)
+	if ok {
+		return all, nil
+	}
+
+	list, err := ssdp.Search(ssdp.All, 3, network) //nolint:mnd
+	if err != nil {
+		return nil, err
+	}
+
+	all = []UPNPDevice{}
+
+	for _, srv := range list {
+		logger.Debugf("found service (%s) at %s", srv.Type, srv.Location)
+
+		desc, err := readDeviceDesc(ctx, srv.Location)
+		if err != nil {
+			logger.Warnf("cannot read description %v", err)
+			continue
+		}
+
+		logger.Debugf("got description %v", desc)
+
+		all = append(all, UPNPDevice{srv, desc})
+	}
+
+	return all, nil
+}
+
+type deviceDesc struct {
+	XMLName     xml.Name `xml:"root"`
+	SpecVersion struct {
+		Major int `xml:"major"`
+		Minor int `xml:"minor"`
+	} `xml:"specVersion"`
+	Device struct {
+		Manufacturer string `xml:"manufacturer"`
+		ModelName    string `xml:"modelName"`
+		SerialNumber string `xml:"serialNumber"`
+	} `xml:"device"`
+}
+
+func readDeviceDesc(ctx context.Context, url string) (*deviceDesc, error) {
+	cli := &http.Client{}
```

Comment:
[nit] this will create a new http client per call. Should be okay though since this code path is run at most once per (re)configure?

Reply:
```
// near top of file
const httpTimeout = 10 * time.Second

// blah blah blah

client := &http.Client{
    Timeout: httpTimeout,
}
```

Might also be best practice to add a timeout